src/docker/restorableappengine.cc

```diff
@@ -517,15 +524,89 @@ bool RestorableAppEngine::isAppFetched(const App& app) const {
       break;
     }
 
-    // No need to check hashes of a Merkle tree of each App image
-    // since skopeo does it internally within in the `skopeo copy` command
+    // TODO: we actually should extract a compose file from the app archive, otherwise
+    // it won't be merkle tree like verification since this compose file can be hacked
+    // because it doesn't have any hash to be verified for.
+    // As an alternative we might consider adding a new attribute to an App root manifest that denotes a hash of a
+    // compose file. Or, a compose yaml should be excluded from an App's archive and considered as App config that is
+    // referenced by ["config"]["digest"] element of an App manifest. (currently this digest is equal to the hash of
+    // empty string)
+    const auto compose_file{app_dir / ComposeFile};
+    if (!boost::filesystem::exists(compose_file)) {
+      LOG_DEBUG << app.name << ": missing App compose file: " << compose_file;
+      break;
+    }
+
+    // No need to check hashes of a Merkle tree of each App image since skopeo does it internally within in the `skopeo
+    // copy` command. While the above statement is true there is still a need in traversing App's merkle tree at the
+    // phase of checking whether Apps are in sync with Target because `skopeo copy` is invoked only if the check detects
+    // any missing piece of App. If the given check doesn't detect any missing part of App then the fetch&install phase
+    // never happens and `skopeo copy` is never invoked. So, let's do in areAppImagesFetched() method.
 
-    res = true;
+    res = areAppImagesFetched(app);
   } while (false);
 
   return res;
 }
 
+bool RestorableAppEngine::areAppImagesFetched(const App& app) const {
+  const Uri uri{Uri::parseUri(app.uri)};
+  const auto app_dir{apps_root_ / uri.app / uri.digest.hash()};
+  const auto compose_file{app_dir / ComposeFile};
+
+  ComposeInfo compose{compose_file.string()};
+  for (const auto& service : compose.getServices()) {
+    const auto image = compose.getImage(service);
+    const Uri image_uri{Uri::parseUri(image)};
+    const auto image_root{app_dir / "images" / image_uri.registryHostname / image_uri.repo / image_uri.digest.hash()};
+
+    const auto index_manifest{image_root / "index.json"};
+    if (!boost::filesystem::exists(index_manifest)) {
+      LOG_DEBUG << app.name << ": missing index manifest of App image; image: " << image
+                << "; index: " << index_manifest;
+      return false;
+    }
+
+    const auto manifest_desc{Utils::parseJSONFile(index_manifest)};
+    HashedDigest manifest_digest{manifest_desc["manifests"][0]["digest"].asString()};
+
+    const auto manifest_file{blobs_root_ / "sha256" / manifest_digest.hash()};
+    if (!boost::filesystem::exists(manifest_file)) {
+      LOG_DEBUG << app.name << ": missing App image manifest; image: " << image << "; manifest: " << manifest_file;
```

Comment:
is this really LOG_DEBUG - almost feels like a LOG_WARN

Reply:
I don't have strong opinion about it. I put LOG_DEBUG because it might be just a regular/normal case if a download is interrupted.